spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/datastore/GcpDatastoreAutoConfiguration.java

```diff
@@ -102,20 +106,31 @@
 
 	@Bean
 	@ConditionalOnMissingBean
-	public Datastore datastore() {
-		DatastoreOptions.Builder builder = DatastoreOptions.newBuilder()
-				.setProjectId(this.projectId)
-				.setHeaderProvider(new UserAgentHeaderProvider(this.getClass()))
-				.setCredentials(this.credentials);
-		if (this.namespace != null) {
-			builder.setNamespace(this.namespace);
-		}
+	public DatastoreNamespaceProvider namespaceProvider() {
+		return () -> this.namespace;
+	}
 
-		if (this.host != null) {
-			builder.setHost(this.host);
+	@Bean
+	@ConditionalOnMissingBean(value = Datastore.class, parameterizedContainer = Supplier.class)
+	public Supplier<Datastore> datastoreSupplier(DatastoreNamespaceProvider namespaceProvider,
+			Optional<Datastore> datastore) {
```

Comment:
I don't think `Optional` works the way you expect it here.
I think you need to use `org.springframework.beans.factory.ObjectProvider` instead.

Reply:
Yes I can change, but I saw this post from STephan https://stackoverflow.com/questions/49044770/change-in-how-spring-5-handles-null-beans

He seems to be saying both can be used the same. I have some tests do confirm it's working as expected with Optional.

Reply:
Just curious, what's the difference between the two?